src/main/java/com/google/sps/data/EventStorage.java

```diff
@@ -19,17 +19,92 @@
+    Query query = new Query("Event");
+
+    if (searchDuration != null && !searchDuration.isEmpty()) {
+      Long searchDurationLong = Long.parseLong(searchDuration);
+      if (searchDurationLong != null) {
+        Filter durationFilter =
+        new FilterPredicate("duration", FilterOperator.LESS_THAN_OR_EQUAL, searchDurationLong);
+        query = query.setFilter(durationFilter);
+      }
+    }
+
+    if (searchLocation != null && !searchLocation.isEmpty() && !searchLocation.equals("Everywhere")) {
+      Filter locationFilter =
+      new FilterPredicate("location", FilterOperator.EQUAL, searchLocation);
+      query = query.setFilter(locationFilter);
+    }
+ 
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    List<Event> events = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      String title = (String) entity.getProperty("title");
+      String description = (String) entity.getProperty("description");
+ 
+      if (search == null || search.isEmpty() || EventStorage.isTextMatch(search, title) || EventStorage.isTextMatch(search, description)) {
+        // TODO: Likely this is an incorrect way to get id
+        String id = (String) entity.getKey().getAppId();
+        String category = (String) entity.getProperty("category");
+
+        if (searchCategory == null || searchCategory.equals("all") || category.equals(searchCategory)) {
+          String gcalendarId = (String) entity.getProperty("gcalendar-id");
+          List<String> tags = (List<String>) entity.getProperty("tags");
+          String dateTimeRangeJson = (String) entity.getProperty("date-time-range");
+          // TODO: Get actual data structure from Json
+          DateTimeRange dateTimeRange = null;
+          Long duration = (Long) entity.getProperty("duration");
+          String location = (String) entity.getProperty("location");
+          List<String> links = (List<String>) entity.getProperty("links");
+          String fieldsJson = (String) entity.getProperty("fields");
+          Map<String, String> fields = new Gson().fromJson(
+          fieldsJson, new TypeToken<HashMap<String, String>>() {}.getType()
+          );
+          String ownerId = (String) entity.getProperty("owner");
+          List<String> invitedParticipantsId = (List<String>) entity.getProperty("invited-users");
+          List<String> joinedParticipantsId = (List<String>) entity.getProperty("joined-users");
+          List<String> declinedParticipantsId = (List<String>) entity.getProperty("declined-users");
+          Event event = new Event(id, gcalendarId, title, description, category, tags, 
+               dateTimeRange, duration, location, links, fields, ownerId, invitedParticipantsId, 
+               joinedParticipantsId, declinedParticipantsId);
+          events.add(event);
```

Comment:
I would suggest to create a factory in the Event class to avoid doing the conversion between entity and event in different places. 
Factory means:
```
class Event {
 public static Event fromDatestoreEntitty(Entity entity) {...}
}
```

Then you can just call `Event event = Event.fromDatastoreEntity(entity);` here.